### ISSUE TITLE
RFC: Support Sass-like variables 

### ISSUE THREAD

**Issue by sonaye** [styled-components/styled-components#1747]
How do you guys feel about this?

```js
// current
const Card = styled.div`
  background: ${props => props.color};
  height: ${props => props.size};
  width: ${props => props.size};
`;

// proposed
const Card = styled.div`
  background: $color;
  height: $size;
  width: $size;
`;

// nested props
const Card = styled.div`
  background: $theme.primary.color;
  height: $size;
  width: $size;
`;

// manipulating props
const LongCard = styled.div`
  background: $theme.primary.color;
  height: ${props => props.size * 1.5}; // sorry, you have to do it this way, any ideas?
  width: $size;
`;
```

**Comment by sonaye** [comment]
Say hello to my little friend!

```js
// `foo: $bar;` wont really work
// we need to properly identify where the variable starts/ends

// ends with a semicolon
foo: $bar; // works
foo: $bar;px; // meh
foo: rgb($r;, $g;, $b;); // 💩

// ends with a space
foo: $bar ; // prettier removes space, fails
foo: $bar px; // meh
foo: rgb($r , $g , $b ); // prettier removes spaces, fails

// parentheses
foo: $(bar); // not bad
foo: $(bar)px; // not bad
foo: rgb($(r), $(g), $(b)); // not that bad
```

```js
// takes a template string style and returns it after evaluating it's props
const parse = props => str => {
  // simple regex, not my strongest skill, feel free to improve
  const regex = /\$\((.+)\)/g;

  // not sure if we need to iterate with str.map()
  let style = str[0];

  // access style instance
  style = style.replace(regex, (match, value) => {
    // split it by a dot to allow nested props (if any)
    let path = value.split('.');

    // evaluate props
    path = path.reduce((obj, prop) => {
      // prop exists
      if (obj) return obj[prop];

      return '';
    }, props);

    return path;
  });

  return style;
};

// test
it('works', () => {
  // assuming props is accessible when calling styled()
  const props = {
    color: 'red',
    size: 64,
    theme: {
      largeText: 36
    },
    deeply: {
      nested: {
        objects: {
          work: 'yes!'
        }
      }
    },
    multi: 'lines'
  };

  const parsed = parse(props)`
    background: $(color);
    height: $(size)px;
    width: $(size)px;
    do: $(deeply.nested.objects.work);
    has:
      $(multi);
  `;

  const expected = `
    background: ${props.color};
    height: ${props.size}px;
    width: ${props.size}px;
    do: yes!;
    has:
      lines;
  `;

  expect(parsed).toBe(expected); // passes
});
```

**Comment by sonaye** [comment]
Will see if it's pluggable to `sc` without causing a test failure and draft a PR.

**Comment by kitten** [comment]
We've basically rejected very similar proposals before on the simple basis that we're trying not to introduce any special syntax and I hope that this makes sufficient sense 👍 

https://github.com/styled-components/styled-components/issues/590
https://github.com/styled-components/styled-components/issues/560

I think there were more issues like these.

Anyway, to propose an alternative, this would be well suited for a babel plugin that doesn't even require any hacky plugin (we don't support plugins), or a fork (worst-case scenario in my opinion, especially for just a feature). As long as you can write the regex and split the string you can simply transpile your strings into interpolations during transpilation time.

**Comment by sonaye** [comment]
Thanks for the reply. I completely understand why you would turn down such proposals that attempt to introduce new syntax. I actually think that you are doing a great job making things light and intuitive, but let make the case for this one -as if I wasn't the one behind it- at least.

I use `sc` because it's a tool that I found to make me more productive while working on user interfaces. When it comes to being productive, I think we should always exhaust that area with investigation, anything that can help developers build something better in a more efficient way is probably a career-devoting worthy. I don't use React just because it's a great tool, I don't use styled components just because it's a great tool, I use both because I love using them. There is a significant difference here, there is an abundance of great software, but finding software that you enjoy is rare. The reason these tools are endorsed by the community is that they are supported by the community itself, in a way it shapes them to fit its needs.

First, I would like to point out that you do generously take into consideration feedback from your users, even if it was syntax related. An example of that would be supporting `// inline comments` in the styles, because it makes out life slightly a bit easier. I would argue that CSS variables would too, and they fit the spirit of React (style as a function of props), and the spirit of styled components (superpowers for CSS). Second, we are not introducing any complex syntax in this case unlike in the proposals you referenced -and I'm sure in many others-. This is new addition `()`, literarily, there is not learning curve, noting complicated, but it comes with great power behind it.

So why did I propose this in the first place? simple, because I disliked the way I have to write `${props => props.foo}` every time, and I write a lot of components. This has nothing with `sc`, this is JS, but we are allowed to hate on JS aren't we? I saw an area where things can be improved, I saw an area that wasn't covered by `sc` (Sass variables), and voila! I really think developers would love it and start using it everywhere, all the time. Maybe I'm wrong, but for now I'd have to trust my understanding of how can a developer be lazy and productive at the same time.

Maybe the proposal here isn't perfect -and it doesn't have to be-, but if you are open to this, you will get my commitment to workout an ideal solution as much as possible. I just [found out](https://lodash.com/docs/4.17.10#get) that there is a method in `lodash` for accessing objects by a string path for example, I would appreciate a chance to continue working on this, or at least keeping the discussion open.

**Comment by sonaye** [comment]
I would appreciate it if you can mention why you think this particular syntax is confusing. Also, would you say that supporting sass-like variables is something that's maybe worth looking into in the future?

Can you provide me with some babel-plugin references for styled-components @mxstbr? maybe it's time for me to learn how to babel.

Link to repo of implementation with tests https://github.com/sonaye/styled-components-variables.

**Comment by kitten** [comment]
It's not the syntax inherently that is confusing, but it blurs the line of where styled-components starts and ends. Right now it's quite easy for us to say that it does auto-prefixing and handles nested CSS. Most of this is without consequence to anyone who knows CSS, no matter whether they're getting started or have been writing it for a while.

A new syntax simply does not belong in a tool that aims to make styled-components a consequent, minimal idea. It might be that this library seems large after a time of growth and refinement, but the core idea is simple. Explaining unique features is burdensome for beginners although it's not without merits. At the same time we'd then have to determine what ideas do belong in the library. The core and its message would change. This is mainly why we've avoided extendability and custom syntax so far.

No promises but if we ever arrive at a whole CSS-in-JS ecosystem this could become much simpler and an easier question to answer

**Comment by sonaye** [comment]
@kitten is this snippet enough to detect all styled components?

https://github.com/styled-components/babel-plugin-styled-components/blob/master/src/utils/detectors.js#L36-L69

Then we can do something like:
```js
export default function({ types: t }) {
  return {
    visitor: {
      TaggedTemplateExpression(path, state) {
        if (isStyled(path, state)) {
          // do stuff
        }
      }
    }
  };
}
```

**Comment by sonaye** [comment]
Got it [working](https://astexplorer.net/#/gist/4fe431202988664af8dcd57e153ffe71/d9020c4becff42b26801723ab7d0bae7f5d95567), not sure if it's enough though.

```js
export default function({ types: t }) {
  // source: https://github.com/styled-components/babel-plugin-styled-components/blob/master/src/utils/detectors.js
  const importLocalName = (name, state) => {
    let localName = name === 'default' ? 'styled' : name;

    state.file.path.traverse({
      ImportDeclaration: {
        exit(path) {
          const { node } = path;

          if (node.source.value === 'styled-components') {
            for (const specifier of path.get('specifiers')) {
              if (specifier.isImportDefaultSpecifier()) {
                localName = specifier.node.local.name;
              }

              if (
                specifier.isImportSpecifier() &&
                specifier.node.imported.name === name
              ) {
                localName = specifier.node.local.name;
              }

              if (specifier.isImportNamespaceSpecifier()) {
                localName = specifier.node.local.name;
              }
            }
          }
        }
      }
    });

    return localName;
  };

  const isStyled = (tag, state) => {
    if (
      t.isCallExpression(tag) &&
      t.isMemberExpression(tag.callee) &&
      tag.callee.property.name !== 'default' /** ignore default for #93 below */
    ) {
      // styled.something()
      return isStyled(tag.callee.object, state);
    } else {
      return (
        (t.isMemberExpression(tag) &&
          tag.object.name === importLocalName('default', state)) ||
        (t.isCallExpression(tag) &&
          tag.callee.name === importLocalName('default', state)) ||
        /**
         * #93 Support require()
         * styled-components might be imported using a require()
         * call and assigned to a variable of any name.
         * - styled.default.div``
         * - styled.default.something()
         */
        (state.styledRequired &&
          t.isMemberExpression(tag) &&
          t.isMemberExpression(tag.object) &&
          tag.object.property.name === 'default' &&
          tag.object.object.name === state.styledRequired) ||
        (state.styledRequired &&
          t.isCallExpression(tag) &&
          t.isMemberExpression(tag.callee) &&
          tag.callee.property.name === 'default' &&
          tag.callee.object.name === state.styledRequired)
      );
    }
  };
  // ..

  return {
    visitor: {
      TaggedTemplateExpression(path, state) {
        if (isStyled(path.node.tag, state)) {
          let { value } = path.node.quasi.quasis[0];
          value.raw = value.raw.toUpperCase();
        }
      }
    }
  };
}
```

**Comment by probablyup** [comment]
I don't think this is a good idea. Part of the reason why explicit interpolations are better for this sort of thing is syntax issues and missing variables are easy to catch with a linter since at the end of the day, it's vanilla JS that most tooling understands.
